ruby/lib/gherkin/query.rb

```diff
@@ -20,34 +27,42 @@ def update_feature(feature)
       store_nodes_location(feature.tags)
 
       feature.children.each do |child|
-        update_rule(child.rule) if child.rule
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_rule(feature, child.rule) if child.rule
+        update_background(feature, child.background) if child.background
+        update_scenario(feature, child.rule, child.scenario) if child.scenario
       end
     end
 
-    def update_rule(rule)
+    def update_rule(feature, rule)
       return if rule.nil?
       store_nodes_location(rule.tags)
-
       rule.children.each do |child|
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_background(rule, child.background) if child.background
+        update_scenario(feature, rule, child.scenario) if child.scenario
       end
     end
 
-    def update_background(background)
+    def update_background(parent, background)
```

Comment:
Think this should be `rule` based on how you've done it in the other methods.

Reply:
This is the more abstract `parent`, because sometimes the direct parent of a Background is a Feature, and other times its a Rule. For example:
```
Feature: Cool feature is cool
Scenario: I become cool by using cool feature
```

```
Feature: Cool feature is cool
Rule: Only uncool people can become cool
Scenario: Already cool person cannot use cool feature
```